### PR TITLE
[Merged by Bors] - Fix spawning empty bundles

### LINKED ISSUE #6226: `World.spawn(())` panics with index out of bounds
> `world.spawn(())` seems to panic with "index out of bounds: the len is 0 but the index is 0". But `commands.spawn(())` seems fine.

_Originally posted by @Zeenobit in https://github.com/bevyengine/bevy/issues/6186#issuecomment-1273856220_

This bug was introduced with @cart's #6054. I think it's important that we fix this for 0.9 to avoid introducing nasty crashes in people's code.

### PR DESCRIPTION
**Pull request by cart**
# Objective

Alternative to #6424 
Fixes #6226

Fixes spawning empty bundles

## Solution

Add `BundleComponentStatus` trait and implement it for `AddBundle` and a new `SpawnBundleStatus` type (which always returns an Added status). `write_components` is now generic on `BundleComponentStatus` instead of taking `AddBundle` directly. This means BundleSpawner can now avoid needing AddBundle from the Empty archetype, which means BundleSpawner no longer needs a reference to the original archetype.

In theory this cuts down on the work done in `write_components` when spawning, but I'm seeing no change in the spawn benchmarks.

### CODE REVIEW

File: crates/bevy_ecs/src/archetype.rs [hunk 1]
```diff
@@ -35,6 +35,7 @@ impl ArchetypeId {
     }
 }
 
+#[derive(Copy, Clone)]
 pub(crate) enum ComponentStatus {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/archetype.rs:39]: Basic docs here to help explain how this new trait is being used?

File: crates/bevy_ecs/src/bundle.rs [hunk 2]
```diff
@@ -360,11 +360,11 @@ impl BundleInfo {
     /// `entity`, `bundle` must match this [`BundleInfo`]'s type
     #[inline]
     #[allow(clippy::too_many_arguments)]
-    unsafe fn write_components<T: Bundle>(
+    unsafe fn write_components<T: Bundle, S: BundleComponentStatus>(
         &self,
         table: &mut Table,
         sparse_sets: &mut SparseSets,
-        add_bundle: &AddBundle,
+        bundle_component_status: &S,
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/bundle.rs:367]: The relationship between `T` and `S` is hard to follow here. I'm unsure about whether this should be documented on this method, BundleComponentStatus, or both.

File: crates/bevy_ecs/src/archetype.rs [hunk 3]
```diff
@@ -45,6 +46,32 @@ pub struct AddBundle {
...
+        *self.bundle_status.get_unchecked(index)
+    }
+}
+
+pub(crate) struct SpawnBundleStatus;
+
+impl BundleComponentStatus for SpawnBundleStatus {
+    #[inline]
+    unsafe fn get_status(&self, _index: usize) -> ComponentStatus {
+        ComponentStatus::Added
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/archetype.rs:71]:
```suggestion
       // Components added during a spawn_bundle call are always treated as added
        ComponentStatus::Added
```

File: crates/bevy_ecs/src/archetype.rs (hunk 1 above)
Comment (cart) [reply crates/bevy_ecs/src/archetype.rs:39]: I'm slightly confused by this comment. This is just making an existing internal type Copy/Clone.

File: crates/bevy_ecs/src/archetype.rs (hunk 3 above)
Comment (cart) [reply crates/bevy_ecs/src/archetype.rs:74]: Good call! I applied this manually as the indentation was off in this suggestion.

File: crates/bevy_ecs/src/bundle.rs (hunk 2 above)
Comment (cart) [reply crates/bevy_ecs/src/bundle.rs:380]: I've added clarifying docs on the requirements and behavior here.

File: crates/bevy_ecs/src/archetype.rs (hunk 1 above)
Comment (cart) [reply crates/bevy_ecs/src/archetype.rs:39]: I'm guessing you meant this comment to be on BundleComponentStatus. I've added clarifying documentation to the behavior and role of this trait.
